src/compiler/checker.ts

```diff
@@ -16024,9 +16024,37 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         // where different generations of the same type parameter are in scope). This leads to a lot of new type
         // identities, and potentially a lot of work comparing those identities, so here we create an instantiation
         // that uses the original type identities for all unconstrained type parameters.
+        //
+        // Whenever a type parameter is replaced with its original other type parameters must be cloned
+        // and appropriate mapper must be associated with them. This is necessary to ensure that their constraints
+        // refer to those replacement type parameters.

```

Comment:
A simpler fix is possible (see [the first commit in this PR](https://github.com/microsoft/TypeScript/pull/59409/commits/33915949aa3a456b4659bfdb0e5e6b7c2b855fbe)) but it makes the type instantiation to go up in one of the tests.

But since it turns out that some type parameters have to be cloned after some type parameters are replaced... it feels like balancing what cases are more common etc. Effectively this tries to reduce some type identitities but some new ones have to be created. Intuitively it feels like this still is net positive though.